File: java/revenj-servlet/src/main/java/org/revenj/serialization/WireSerialization.java
package org.revenj.serialization;

import org.revenj.Utils;

import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.lang.reflect.Type;
import java.util.Optional;

public interface WireSerialization {
	String serialize(Object value, OutputStream stream, String accept) throws IOException;

	default ByteArrayOutputStream serialize(Object value, String accept) throws IOException {
		ByteArrayOutputStream os = new ByteArrayOutputStream();
		serialize(value, os, accept);
		return os;
	}

	Object deserialize(Type type, byte[] content, int length, String contentType) throws IOException;

	Object deserialize(Type type, InputStream stream, String contentType) throws IOException;

	@SuppressWarnings("unchecked")
	default <T> T deserialize(byte[] content, int length, String contentType, Class<T> manifest) throws IOException {
		return (T) deserialize(manifest, content, length, contentType);
	}

	@SuppressWarnings("unchecked")
	default <T> T deserialize(byte[] content, String contentType, Class<T> manifest) throws IOException {
		return (T) deserialize(manifest, content, content.length, contentType);
	}

	@SuppressWarnings("unchecked")
	default <T> T deserialize(InputStream stream, String contentType, Class<T> manifest) throws IOException {
		return (T) deserialize(manifest, stream, contentType);
	}

	@SuppressWarnings("unchecked")
	default <T> T deserialize(byte[] content, int length, String contentType, Class<T> container, Type argument, Type... arguments) throws IOException {
		return (T) deserialize(Utils.makeGenericType(container, argument, arguments), content, length, contentType);
	}

	@SuppressWarnings("unchecked")
	default <T> T deserialize(InputStream stream, String contentType, Class<T> container, Type argument, Type... arguments) throws IOException {
		return (T) deserialize(Utils.makeGenericType(container, argument, arguments), stream, contentType);
	}

	<TFormat> Optional<Serialization<TFormat>> find(Class<TFormat> format);
}

File: scala/revenj-akka/src/main/scala/net/revenj/server/WireSerialization.scala
package net.revenj.server

import java.io.{ByteArrayOutputStream, InputStream, OutputStream}
import java.lang.reflect.Type

import net.revenj.Utils
import net.revenj.serialization.Serialization

import scala.util.{Failure, Success, Try}
import scala.reflect.runtime.universe._

trait WireSerialization {
  def serialize(value: Any, stream: OutputStream, contentType: String, manifest: Type): Try[String]

  def serialize(value: Any, stream: OutputStream, contentType: String): Try[String] = {
    if (value != null) serialize(value, stream, contentType, value.getClass)
    else serialize(null, stream, contentType, classOf[AnyRef])
  }

  def serialize(value: Any, accept: String): Try[ByteArrayOutputStream] = {
    Try {
      val os = new ByteArrayOutputStream
      serialize(value, os, accept)
      os
    }
  }

  def deserialize(manifest: Type, content: Array[Byte], length: Int, contentType: String): Try[Any]

  def deserialize(manifest: Type, stream: InputStream, contentType: String): Try[Any]

  def deserialize[T](manifest: Class[T], stream: InputStream, contentType: String): Try[T] = {
    val signature: Type = manifest
    deserialize(signature, stream, contentType) match {
      case Success(res: T@unchecked) => Success(res)
      case Failure(ex) => Failure(ex)
    }
  }

  def deserialize[T: TypeTag](content: Array[Byte], length: Int, contentType: String): Try[T]

  def deserialize[T: TypeTag](content: Array[Byte], contentType: String): Try[T] = {
    deserialize[T](content, content.length, contentType)
  }

  def deserialize[T: TypeTag](stream: InputStream, contentType: String): Try[T]

  def deserialize[T](content: Array[Byte], length: Int, contentType: String, container: Class[T], argument: Type, arguments: Type*): Try[T] = {
    deserialize(Utils.makeGenericType(container, argument, arguments: _*), content, length, contentType) match {
      case Success(res: T@unchecked) => Success(res)
      case Failure(ex) => Failure(ex)
    }
  }

  def deserialize[T](stream: InputStream, contentType: String, container: Class[T], argument: Type, arguments: Type*): Try[T] = {
    deserialize(Utils.makeGenericType(container, argument, arguments: _*), stream, contentType) match {
      case Success(res: T@unchecked) => Success(res)
      case Failure(ex) => Failure(ex)
    }
  }

  def find[TFormat: TypeTag](): Option[Serialization[TFormat]]
}
